lib/pg.provider.js

```diff
@@ -0,0 +1,286 @@
+'use strict';
+
+const path = require('path');
+
+const metaschema = require('metaschema');
+const pg = require('pg');
+
+const { GSError, codes: errorCodes } = require('./errors');
+const { StorageProvider } = require('./provider');
+const {
+  generateQueryParams,
+  escapeIdentifier,
+  buildWhere,
+} = require('./pg.utils');
+const { PostgresCursor } = require('./pg.cursor');
+
+const recreateIdTrigger = Symbol('recreateIdTrigger');
+
+class PostgresProvider extends StorageProvider {
+  // Create PostgresProvider
+  //   gs - globalstorage instance
+  constructor(gs) {
+    super(gs);
+
+    this.pool = null;
+  }
+
+  // Open PostgresProvider
+  //   options - object, passed to pg
+  //   callback - function(err, StorageProvider)
+  open(options, callback) {
+    metaschema.load(path.join(__dirname, '../schemas/system'), err => {
+      if (err) {
+        callback(err);
+        return;
+      }
+
+      this.pool = new pg.Pool(options);
+      callback(null, this);
+    });
+  }
+
+  // Close PostgresProvider
+  //   callback - function(err)
+  close(callback) {
+    if (!this.pool) {
+      callback();
+      return;
+    }
+
+    this.pool.end(() => {
+      this.pool = null;
+      callback();
+    });
+  }
+
+  [recreateIdTrigger](maxIdCount, refillPercent, callback) {
+    this.pool.query('DROP TRIGGER IF EXISTS idgen ON "Identifier"', err => {
+      if (err) {
+        callback(err);
+      }
+
+      this.pool.query(
+        'CREATE TRIGGER idgen BEFORE UPDATE ON "Identifier"' +
+        'FOR EACH STATEMENT EXECUTE FUNCTION idgen($1, $2, $3, $4)',
+        [maxIdCount, refillPercent, this.gs.serverId, this.gs.serverIdMask],
+        err => {
+          callback(err);
+        }
+      );
+    });
+  }
+
+  // Generate globally unique id
+  //   callback - function(err, id)
+  takeId(callback) {
+    const takeIdQuery = 'UPDATE "Identifier"' +
+      ' SET "Status" = \'Init\', "Change" = CURRENT_TIMESTAMP' +
+      ' WHERE "Id" = (SELECT min("Id")' +
+      ' FROM "Identifier" WHERE "Status" = \'Prealloc\'' +
+      ') RETURNING "Id"';
+    this.pool.query(takeIdQuery, (err, res) => {
+      if (err) {
+        callback(err);
+        return;
+      }
+
+      if (res.rowCount === 0) {
+        callback(new GSError(
+          errorCodes.NOT_FOUND,
+          'Cannot get Id to use for object creation'
+        ));
+        return;
+      }
+
+      callback(null, res.rows[0].Id);
+    });
+  }
+
+  getCategoryById(id, callback) {
+    const categoryQuery = 'SELECT "Category" FROM "Identifier" WHERE "Id" = $1';
+    this.pool.query(categoryQuery, [id], (err, res) => {
+      if (err) {
+        callback(err);
+        return;
+      }
+      if (res.rowCount === 0) {
+        callback(new GSError(
+          errorCodes.NOT_FOUND,
+          `No object with Id ${id} available`
+        ));
+        return;
+      }
+      const { Category } = res.rows[0];
```

Comment:
What about `const [{ Category }] = res.rows;` ?

Reply:
looks cryptic, though I'm fine with either.

Reply:
It doesn't help readability, so I'd prefer to leave it as it is now.